python/GafferSceneUI/InstancerUI.py

```diff
@@ -306,6 +640,142 @@
+			"plugValueWidget:type", "GafferSceneUI.InstancerUI._ContextVariableListWidget",
+		],
+
+		"contextVariables.*" : [
+			"deletable", True
+		],
+
+		"contextVariables.*.name" : [
+			"description",
+			"""
+			Name of the primitive variable to read.  The same name will be used for the context variables created,
+			which are used to used to evaluate the protoypes plug.
+			differently depending on the source point.  Supports quantization to avoid re-evaluating the
+			prototypes scene too many times.
+			""",
+		],
+
+		"contextVariables.*.enabled" : [
+			"description",
+			"""
+			Put this variable in the context for the prototypes plug.
+			""",
+		],
+
+		"contextVariables.*.quantize" : [
+			"description",
+			"""
+			Quantizing to a large interval reduces the number of variations created.  For example, if the primvar varies from 0 to 1, and you quantize to 0.2, then only 6 unique variations will be created, even if there are millions of instances.  This dramatically improves performance, but if you need to see more continuous changes in the primvar values, you will need to reduce quantize, or in extreme cases where you need full accuracy and don't care about performance, set it to 0.
+			""",
+		],
+
+		"timeOffset" : [
+			"description",
+			"""
+			Modify the current time when evaluating the prototype scene, by adding a primvar.
+			""",
+			"layout:section", "Context Variations",
+			"plugValueWidget:type", "GafferSceneUI.InstancerUI._TimeOffsetContextVariableWidget",
+		],
+		"timeOffset.name" : [
+			"description",
+			"""
+			Name of a primitive variable to add to the time.  Must be a float or int primvar.  It will
+			be treated as a number of frames, and can be negative or positive to adjust time forward or back.
+			""",
+		],
+		"timeOffset.enabled" : [
+			"description",
+			"""
+			Modify the current time when evaluating the prototypes plug.
+			""",
+		],
+		"timeOffset.quantize" : [
+			"description",
+			"""
+			Quantize the variable value before adding it to the time.  Quantizing to a large interval reduces the number of variations created.  For example, if the primvar varies from 0 to 1, and you quantize to 0.2, then only 6 unique variations will be created, even if there are millions of instances.  This dramatically improves performance, but if you need to see more continuous changes in the primvar values, you will need to reduce quantize, or in extreme cases where you need full accuracy and don't care about performance, set it to 0.
+			""",
+		],
+
+		"variations" : [
+			"description",
+			"""
+			Varying contexts requires extra evaluations of the `prototypes` scene.  These counts
+			show the number of unique values required by each variation source, and the total.
+			Increasing the number of variations required can dramatically increase the cost of
+			the Instancer.  Note that variations are measured across all locations in the scene
+			where the instancer is filtered.
+			""",
```

Comment:
That definitely needs to be documented somewhere, but it feels a little bit weird to put it in the tooltip which pops up in the UI - I picture the primary consumer of that to be a user who wants to know "What does this display in the UI mean?", where the widget has already taken care of unpacking the correct channel - and talking about the internal data format seems like a distraction from something we really want to emphasize to users as clearly as possible.

I guess we do run into a tension here with the convention that Gaffer nodes don't really distinguish between the interface for the UI, and for scripting ... my instinct would be to put programmer facing documentation in the header or something, but I guess  that's non-standard.

Reply:
I think it's certainly OK to put programmer-only stuff in the header, but only where that information isn't useful as a user or a python scripter, and doesn't belong in the Node Reference in the docs. Here we're documenting the data format being output by a public plug, so it definitely belongs in the node reference.

There is still a tension between the Node Reference and the tooltips though. We could perhaps introduce a separate `tooltip` metadata that overrides `description` for this context? Or perhaps because the UI is so specific in this case anyway, you could set the tooltip manually?

Reply:
Addressed with kludge of scripting related documentation in the "description", with UI toolTip override so that users see something more user focussed.